File: iteration2/src/CSE3063F22P1_GRP9/InputJSON.java
package CSE3063F22P1_GRP9;

import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.Reader;
import java.util.ArrayList;
import java.util.Iterator;

import org.apache.log4j.Logger;
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;

public class InputJSON {
	final Logger logger = Logger.getLogger(InputJSON.class);
	private JSONObject input;
	private ArrayList<ArrayList<Course>> courses;
	public InputJSON(){
		try {
			Reader reader = new FileReader("parameters.json");
			JSONParser parser = new JSONParser();
			this.input = (JSONObject) parser.parse(reader);
			courses = new ArrayList<ArrayList<Course>>();
		}catch(FileNotFoundException e) {
			logger.error("parameters.json file isn't found.");
			System.exit(1);
		}catch(Exception e) {
			logger.error("parameters.json file found, there is another issue:");
			e.printStackTrace();
			System.exit(2);
		}
		readCourses();
	}

	private void readCourses(){
		JSONObject coursesJSON = (JSONObject) input.get("courses");
		for(int i = 1;i<9;i++) {
			ArrayList<Course> semesterCourses = new ArrayList<Course>();
			JSONArray semester = (JSONArray)coursesJSON.get("s"+i);
			Iterator<JSONObject> iterator = semester.iterator();
			while (iterator.hasNext()) {
                Course c = getCourseObject(iterator.next());
                semesterCourses.add(c);
            }
			courses.add(semesterCourses);	
		}
		for(int i = 0;i<courses.size();i++) {
			for(int j = 0;j<courses.get(i).size();j++) {
				Course currCourse = courses.get(i).get(j);
				if(currCourse.getPrerequisiteId()!=null) {;
					currCourse.setPrerequisite(findCourse(currCourse.getPrerequisiteId()));
				}
			}
		}
		logger.info("Courses read from inputJSON.");
	}
	private Course getCourseObject(JSONObject courseJSON){
		Course course = new Course();
		course.setID(courseJSON.get("Course Code").toString());
		course.setName(courseJSON.get("Course Name").toString());
		if(courseJSON.get("Prerequisite")!=null) {
			course.setPrerequisiteId(courseJSON.get("Prerequisite").toString());
		}
		course.setQuota(Integer.parseInt(courseJSON.get("Quota").toString()));
		course.setCredit(Integer.parseInt(courseJSON.get("Credit").toString()));
		course.setSemester(courseJSON.get("Semester").toString());
		return course;
	}
	
	private Course findCourse(String courseId) {
		for(int i = 0;i<courses.size();i++) {
			for(int j = 0;j<courses.get(i).size();j++) {;
				if(courses.get(i).get(j).getID().equals(courseId))
					logger.info("course found");
					return courses.get(i).get(j);
			}
		}
		return null;
	}

	public ArrayList<String> getFirstNames() {
		ArrayList<String> names = new ArrayList<String>();
		JSONArray namesJSON =(JSONArray)input.get("firstNames");
		Iterator<String> iterator = namesJSON.iterator();
		while (iterator.hasNext()) {
            names.add(iterator.next());
        };
        return names;
	}
	
	public ArrayList<String> getLastNames(){
		ArrayList<String> lastNames = new ArrayList<String>();
		JSONArray lastNamesJSON =(JSONArray)input.get("lastNames");
		Iterator<String> iterator = lastNamesJSON.iterator();
		while (iterator.hasNext()) {
            lastNames.add(iterator.next());
        };
        return lastNames;
	}
	
	public ArrayList<Advisor> getAdvisors(){
		ArrayList<Advisor> advisors = new ArrayList<Advisor>();
		JSONArray advisorsJSON =(JSONArray)input.get("advisors");
		Iterator<JSONObject> iterator = advisorsJSON.iterator();
		while (iterator.hasNext()) {
			JSONObject advisorJSON = (JSONObject)iterator.next();
			String FName = advisorJSON.get("firstName").toString();
			String LName = advisorJSON.get("lastName").toString();
			String Id = advisorJSON.get("id").toString();
			String email = advisorJSON.get("email").toString();
			String office = advisorJSON.get("office").toString();
			Advisor advisor = new Advisor(Id,FName,LName,email,office);
            advisors.add(advisor);
        };
        return advisors;
	};
	
	public ArrayList<ArrayList<Course>> getCourses() {
		return courses;
	}
	
	public String getSemester() {
		return input.get("semester").toString();
	}
	
	public int getNumberOfStudents() {
		return Integer.parseInt(input.get("numberOfStudents").toString());
	}
	
	public double getProbabilityOfPassingCourse() {
		return Double.parseDouble(input.get("probabilityOfPassingCourse").toString());
	}
}

File: iteration3/src/InputJSON.py
import json
from Advisor import Advisor
from Course import Course
from Schedule import Schedule
import logging
import sys

class InputJSON:
    def __init__(self):
        self._logger = logging.getLogger(__name__)
        try:
         file = open('parameters.json',encoding='utf-8')
        except FileNotFoundError:
            self._logger.critical('Parameter file can not found.')
            sys.exit()

        self._input = json.load(file)
        file.close()
        self._courses = []
        self.readCourses()
    
    
    def readCourses(self):
        coursesJSON = self._input['courses']
        for i in range(1,9):
            semesterCourses = []
            semester = coursesJSON['s'+str(i)]
            for course in semester:
                c = self.getCourseObject(course)
                semesterCourses.append(c)
            self._courses.append(semesterCourses)
        for semester in self._courses:
            for course in semester:
                if course.getPrerequisiteId() != None:
                    prereq = self.findCourse(course.getPrerequisiteId())
                    course.setPrerequisite(prereq)
    
    def getCourseObject(self,courseJSON):
        course = Course()
        course.setID(courseJSON['Course Code'])
        course.setName(courseJSON['Course Name'])
        if courseJSON['Prerequisite']!=None:
            course.setPrerequisiteId(courseJSON['Prerequisite'])
        course.setQuota(courseJSON['Quota'])
        course.setCredit(courseJSON['Credit'])
        course.setSemester(courseJSON['Semester'])
        for i in courseJSON['DayTime']:
            day = i['day']
            startingTime = i['time'][0:2]
            course.addSchedule(Schedule(day,startingTime))
        return course
    
    def findCourse(self,prerequisiteId):
        for semester in self._courses:
            for course in semester:
                if course.getID() == prerequisiteId:
                    return course
        return None

    def getAdvisors(self):
        advisors = []
        advisorsJSON = self._input['advisors']
        for i in advisorsJSON:
            fName = i['firstName']
            lName = i['lastName']
            Id = i['id']
            email = i['email']
            office = i['office'] 
            advisor = Advisor(Id,fName,lName,email,office)
            advisors.append(advisor)
        return advisors
        
    def getFirstNames(self):
        return self._input['firstNames']
    
    def getLastNames(self):
        return self._input['lastNames']
    
    def getCourses(self):
        return self._courses
    
    def getSemester(self):
        try:
         semester = self._input['semester']
         if semester != 'Fall' and semester != 'Spring':
            raise Exception('Simulatoin start in Fall or Spring semesters.')
        except Exception as msg:
            self._logger.critical(msg)
            sys.exit()
        return semester

    def setNumberOfStudents(self,number):
        self._input['numberOfStudents'] = number

        
    def getNumberOfStudents(self):
        try:
         student_number = self._input['numberOfStudents']
         if type(student_number) != type(1) or student_number <= 0 :
            raise ValueError('Not a correct type or student number must be greather than 0')
         if student_number > 1000:
            raise Exception(f'{student_number} students too high for simulation. Simulation limitited for up to 1000 students.')
         return self._input['numberOfStudents']

        except ValueError as msg:
            self._logger.critical(msg)
            sys.exit()
        except Exception as msg:
          self._logger.warning(msg)
          self._logger.warning(f'Simulation will continue up until 1000 student.')
          self.setNumberOfStudents(1000)
          return self.getNumberOfStudents()

    def getProbabilityOfPassingCourse(self):
        try:
         prob = float(self._input['probabilityOfPassingCourse'])
         if prob > 1.0 or prob < 0 :
            raise Exception('Probability must be btwn [0,1]')
         return prob
        except ValueError:
         self._logger.critical('Given \'probabilityOfPassingCourse\' is not float type.')
         sys.exit()
        except Exception as msg:
         self._logger.critical(f'{msg}')
         sys.exit()


